File: wwwroot/reporting.js
const apiUrl = 'http://localhost:5281/api/reporting'; // Update this URL to match your backend

async function submitReport() {
    const documentId = document.getElementById('documentId').value;
    const reason = document.getElementById('reason').value;

    if (!documentId || !reason) {
        alert("Please fill in all fields.");
        return;
    }

    const response = await fetch(`${apiUrl}/CreateReport`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({ DocumentId: documentId, Reason: reason }),
    });

    const feedbackElement = document.getElementById('createReportFeedback');
    if (response.ok) {
        const result = await response.json();
        feedbackElement.innerText = "Report submitted successfully! ID: " + result.id;
        document.getElementById('documentId').value = '';
        document.getElementById('reason').value = '';
    } else {
        const error = await response.json();
        feedbackElement.innerText = "Error: " + error.message;
    }
}

async function fetchReports() {
    const response = await fetch(`${apiUrl}/GetAllReports`);

    const feedbackElement = document.getElementById('fetchReportFeedback');
    const reportsList = document.getElementById('reports-list');
    reportsList.innerHTML = ''; // Clear previous results

    if (response.ok) {
        const reports = await response.json();
        reports.forEach(report => {
            const row = `<tr>
                <td>${report.id}</td>
                <td>${report.documentId}</td>
                <td>${report.reason}</td>
                <td>${report.status}</td>
            </tr>`;
            reportsList.insertAdjacentHTML('beforeend', row);
        });
        feedbackElement.innerText = `${reports.length} reports fetched successfully.`;
    } else {
        const error = await response.json();
        feedbackElement.innerText = "Error: " + error.message;
    }
}

async function updateReportStatus() {
    const id = document.getElementById('reportId').value;
    const status = document.getElementById('status').value;

    if (!id || !status) {
        alert("Please fill in all fields.");
        return;
    }

    const response = await fetch(`${apiUrl}/updateStatus/${id}`, {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({ Status: status }),
    });

    const feedbackElement = document.getElementById('updateStatusFeedback');
    if (response.ok) {
        feedbackElement.innerText = "Report status updated successfully!";
        document.getElementById('reportId').value = '';
    } else {
        const error = await response.json();
        feedbackElement.innerText = "Error: " + error.message;
    }
}

async function deleteApprovedReports() {
    const response = await fetch(`${apiUrl}/DeleteApprovedReports`, {
        method: 'DELETE',
    });

    const feedbackElement = document.getElementById('deleteReportFeedback');
    if (response.ok) {
        const result = await response.json();
        feedbackElement.innerText = result.message;
    } else {
        const error = await response.json();
        feedbackElement.innerText = "Error: " + error.message;
    }
}

File: Reporting.cs
using Microsoft.AspNetCore.Mvc;
using MongoDB.Bson;
using MongoDB.Driver;
using System;
using Microsoft.Extensions.Logging;
using System.Linq;
using System.Threading.Tasks;
using System.Text.Json;

namespace ReportManagement.Controllers
{
    /// <summary>
    /// Handles reporting-related operations.
    /// </summary>
    [ApiController]
    [Route("api/[controller]")]
    public class ReportingController : ControllerBase
    {
        private readonly IMongoCollection<ReportDto> _reportCollection;
        private readonly ILogger<ReportingController> _logger;

        /// <summary>
        /// Initializes a new instance of the <see cref="ReportingController"/> class.
        /// </summary>
        /// <param name="database">The MongoDB database instance.</param>
        public ReportingController(IMongoDatabase database)
        {
            _reportCollection = database.GetCollection<ReportDto>("Reports");
        }

        [HttpPost("CreateReport")]
        public async Task<IActionResult> SubmitReport([FromBody] CreateReportDto newReport)
        {
            try
            {
                _logger.LogInformation($"Received report: DocumentId={newReport?.DocumentId}, Reason={newReport?.Reason}");

                // Validate the input
                if (newReport == null || string.IsNullOrEmpty(newReport.DocumentId) || string.IsNullOrEmpty(newReport.Reason))
                {
                    _logger.LogWarning("Invalid report data received");
                    return BadRequest(new { message = "Please provide all required information (DocumentId and Reason)." });
                }

                // Validate DocumentId format (MongoDB ObjectId)
                if (!ObjectId.TryParse(newReport.DocumentId, out _))
                {
                    _logger.LogWarning($"Invalid DocumentId format: {newReport.DocumentId}");
                    return BadRequest(new { message = "Invalid DocumentId format. It should be a 24-character hexadecimal string." });
                }

                // Create the report DTO
                var report = new ReportDto
                {
                    Id = ObjectId.GenerateNewId().ToString(),
                    DocumentId = newReport.DocumentId,
                    Reason = newReport.Reason,
                    Status = "pending",
                    DateReported = DateTime.UtcNow
                };

                // Insert the report into MongoDB
                _logger.LogInformation($"Inserting report: {JsonSerializer.Serialize(report)}");
                await _reportCollection.InsertOneAsync(report);
                _logger.LogInformation($"Report inserted successfully: Id={report.Id}");

                // Return success response with the created report's ID
                return CreatedAtAction(nameof(GetAllReports), new { id = report.Id }, new { id = report.Id });
            }
            catch (Exception ex)
            {
                // Log the error and return 500 status code
                _logger.LogError(ex, "Error occurred while processing report. Request body: {@newReport}", newReport);
                return StatusCode(500, new { message = "An error occurred while processing your request.", error = ex.Message });
            }
        }


        /// <summary>
        /// Retrieves all reports.
        /// </summary>
        /// <returns>A task representing the asynchronous operation.</returns>
        [HttpGet("GetAllReports")]
        public async Task<IActionResult> GetAllReports()
        {
            var reports = await _reportCollection.Find(r => true).ToListAsync();
            return Ok(reports);
        }

        /// <summary>
        /// Updates the status of an existing report.
        /// </summary>
        /// <param name="id">The ID of the report to update.</param>
        /// <param name="updateDto">The new status data for the report.</param>
        /// <returns>A task representing the asynchronous operation.</returns>
        [HttpPut("updateStatus/{id}")]
        public async Task<IActionResult> UpdateReportStatus(string id, [FromBody] UpdateReportDto updateDto)
        {
            if (updateDto == null)
                return BadRequest(new { message = "Please provide a status to update." });

            // If the status is empty or null, we clear it by setting it to "pending"
            if (string.IsNullOrEmpty(updateDto.Status))
            {
                updateDto.Status = "pending"; // Define what clearing means (set to pending)
            }
            else
            {
                var validStatuses = new[] { "approved", "denied", "pending" }; // Include pending if you want to set it back to this status
                if (!validStatuses.Contains(updateDto.Status.ToLower()))
                    return BadRequest(new { message = "Status must be either 'approved', 'denied', or 'pending'." });
            }

            // Update the status using a case-insensitive comparison
            var update = Builders<ReportDto>.Update.Set(r => r.Status, updateDto.Status);
            var result = await _reportCollection.UpdateOneAsync(
                r => r.Id == id,
                update);

            if (result.ModifiedCount == 0)
                return NotFound(new { message = "Report not found or status unchanged." });

            return Ok(new { message = "Report status updated successfully." });
        }

        /// <summary>
        /// Deletes all approved reports.
        /// </summary>
        /// <returns>A task representing the asynchronous operation.</returns>
        [HttpDelete("DeleteApprovedReports")]
        public async Task<IActionResult> DeleteApprovedReports()
        {
            // Attempt to delete all reports that have the status "approved" (case insensitive)
            var result = await _reportCollection.DeleteManyAsync(r => r.Status.ToLower() == "approved");

            if (result.DeletedCount > 0)
            {
                return Ok(new { message = $"{result.DeletedCount} approved reports deleted." });
            }
            else
            {
                return NotFound(new { message = "No approved reports found to delete." });
            }
        }
    }
}
